### PR TITLE
Generate structs for the return data of abigen-erated contracts

### PR DESCRIPTION
**Pull request by joleeee**
## Motivation
I explain the usecase in #1405, the short version is that you cannot easily decode the return of calls without an instatiated contract (`::new()` with a provider). Since abigen already generates structs representing calls, which AbiEncode into bytes, it only makes sense it should also have structs that AbiDecode the return data into a usable response representation.

## Solution
I hope the code is mostly self explanitory. Since I am new to contributing here, I'd love some feedback on the stuff i did so far.

I'm wondering if you'd want there to be an enum for return structs when there is more than one, like with the calls? I can put it back in but i initially removed it.

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Updated the changelog

### CODE REVIEW

**Comment by gakonst** [comment]
like the direction - could you also add some tests for this?

**Comment by joleeee** [comment]
I made a test now, let me know if it is satisfactory or what else should be tested. Note that this will not generate return bindings for functions that do not return anything.

File: ethers-contract/ethers-contract-abigen/src/contract/methods.rs [hunk 1]
```diff
@@ -135,7 +141,47 @@ impl Context {
...
+    /// Expands to the corresponding struct type based on the inputs of the given function
+    fn expand_return_struct(
+        &self,
+        function: &Function,
+        alias: Option<&MethodAlias>,
+    ) -> Result<TokenStream> {
+        let struct_name = expand_return_struct_name(function, alias);
+        let fields = self.expand_output_params(function)?;
+        // no point in having structs when there is no data returned
+        if function.outputs.len() < 1 {
```
Comment (mattsse) [inline ethers-contract/ethers-contract-abigen/src/contract/methods.rs:153]:
```suggestion
        if function.outputs.is_empty() {
```

**Review by mattsse — CHANGES_REQUESTED** [review]
this only _generates_ structs, right?

this still does not replace the return type of the contract calls, right? those are still tuples?

**Comment by joleeee** [comment]
That is correct, this only adds additional structs, it does not change the call-bindings in any way.

**Review by mattsse — APPROVED** [review]
lgtm

this should not break anything, since this merely generates new types

**Pull request merged**